crates/fluvio-protocol/src/core/bytebuf.rs

```diff
@@ -0,0 +1,301 @@
+use std::io::Error;
+use std::io::ErrorKind;
+
+use bytes::buf::UninitSlice;
+use bytes::{Buf, Bytes, BufMut};
+
+use crate::{Encoder, Decoder, Version};
+
+/// Represnts a SmartModule WASM File bytes.
+///
+/// Provides a `Encoder` and `Decoder` implementation optimized for WASM files
+/// used in SmartModules.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct ByteBuf {
+    inner: Vec<u8>,
+}
+
+impl ByteBuf {
+    pub fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+    }
+
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    pub fn inner(&self) -> Vec<u8> {
+        self.inner.clone()
+    }
+}
+
+impl From<Vec<u8>> for ByteBuf {
+    fn from(bytes: Vec<u8>) -> Self {
+        ByteBuf { inner: bytes }
+    }
+}
+
+impl Decoder for ByteBuf {
+    fn decode<T>(&mut self, src: &mut T, version: Version) -> Result<(), Error>
+    where
+        T: Buf,
+    {
+        let mut len: i32 = 0;
+        len.decode(src, version)?;
+
+        if len < 1 {
+            return Ok(());
+        }
+
+        let mut buf = src.take(len as usize);
+        self.inner
+            .extend_from_slice(&buf.copy_to_bytes(len as usize));
+
+        Ok(())
+    }
+}
+
+impl Encoder for ByteBuf {
+    fn write_size(&self, version: Version) -> usize {
+        self.inner
+            .iter()
+            .fold(4, |sum, val| sum + val.write_size(version))
```

Comment:
Since this is a vec of `u8` I think that we could do size of `i32.write_size(version)` + self.inner.len()

Reply:
Like a charm!